File: C++/SinglyLinkedList.cpp
#include <iostream>
using namespace std;

class Node
{

public:
    int data;
    Node *link = NULL;
};
class SinglyLinkedList
{
public:
    Node *head = new Node();
    Node *tail = new Node();

    int length = 0;
    void push(int val)
    {
        Node *newNode = new Node();
        if (length == 0)
        {
            newNode->data = val;
            head = newNode;
            tail = head;
            length += 1;
        }
        else
        {
            newNode->data = val;
            tail->link = newNode;
            tail = newNode;
            length += 1;
        }
    }
    void pop()
    {
        Node *prev = head;
        while (prev->link->link != NULL)
        {
            prev = prev->link;
        }
        tail = prev;
        tail->link = NULL;
        length -= 1;
        if (length == 0)
        {
            head->link = NULL;
            tail->link = NULL;
        }
    }
    void shift()
    {
        if (length == 0)
        {
            return;
        }
        else
        {
            head = head->link;
            length -= 1;
        }
    }
    void unshift(int val)
    {
        Node *newNode = new Node();
        newNode->data = val;
        newNode->link = head;
        head = newNode;
        length += 1;
    }
    void printList(Node *n)
    {
        while (n != NULL)
        {
            cout << n->data << " ";
            n = n->link;
        }
    }
    void set(int index, int val)
    {

        if (index >= length)
            return;
        int count = 0;
        Node *cur;
        cur = head;
        while (index != count)
        {
            cur = cur->link;
            count += 1;
        }
        cur->data = val;
        return;
    }
    int get(int val)
    {
        if (val >= length)
            return 0;
        int count = 0;
        Node *cur;
        cur = head;
        while (val != count)
        {
            cur = cur->link;
            count += 1;
        }
        return cur->data;
    }
    void remove(int index)
    {
        if (index == 0)
        {
            shift();
            return;
        }
        if (index == length - 1)
        {
            pop();
            return;
        }

        if (index >= length)
            return;
        int count = 0;
        Node *cur;
        cur = head;
        while (index - 1 != count)
        {
            cur = cur->link;
            count += 1;
        }
        cur->link = cur->link->link;
        length -= 1;
    }
    void insert(int index, int value)
    {
        if (index >= length)
            return;
        if (index == 0)
        {
            unshift(value);
            return;
        }
        int count = 0;
        Node *prev = head;
        while (count != index - 1)
        {
            prev = prev->link;
            count += 1;
        }
        Node *newNode = new Node();
        newNode->link = prev->link;
        prev->link = newNode;
        newNode->data = value;
        length += 1;
    }
    void reverse()
    {
        Node *newNode = head;
        head = tail;
        tail = newNode;
        Node *next;
        Node *prev = NULL;
        for (int i = 0; i < length; i++)
        {
            next = newNode->link;
            newNode->link = prev;
            prev = newNode;
            newNode = next;
        }
    }
};
int main()
{
    SinglyLinkedList l1;
    l1.push(10);
    l1.push(11);
    l1.push(13);
    l1.push(23);
    l1.push(15);
    l1.printList(l1.head);
    l1.shift();
    cout << "\n";
    l1.printList(l1.head);
    l1.unshift(99);
    cout << "\n";
    l1.printList(l1.head);
    int give = l1.get(2);
    cout << "\n";
    cout << give;
    l1.set(2, 7);
    cout << "\n";
    l1.printList(l1.head);
    l1.insert(0, 7);
    cout << "\n";
    l1.printList(l1.head);
    cout << "\n";
    l1.remove(0);
    l1.printList(l1.head);
    cout << "\n";
    l1.reverse();
    l1.printList(l1.head);
}

File: JavaScript/SinglyLinkedList.js
class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}
class SinglyLinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }
  push(val) {
    var node = new Node(val);
    if (this.head === null) {
      this.head = node;
      this.tail = node;
      this.length += 1;
    } else {
      this.tail.next = node;
      this.tail = node;
      this.length += 1;
    }
  }
  pop() {
    var prev = this.head;
    var curr;
    while (prev.next.next != null) {
      prev = prev.next;
      var curr = prev.next.next;
    }
    this.tail = prev;
    this.tail.next = null;
    this.length -= 1;
    if (this.length === 0) {
      this.head = null;
      this.tail = null;
      length = 0;
    }
    return curr;
  }
  shift() {
    if (this.head == null) return undefined;
    var curr = this.head.val;
    this.head = this.head.next;
    this.length -= 1;
    return curr;
  }
  unshift(val) {
    var newNode = new Node(val);
    newNode.next = this.head;
    this.head = newNode;
    this.length += 1;
  }
  set(index, val) {
    if (index >= this.length) return false;
    var count = 0;
    var newNode = this.head;
    while (count != index) {
      newNode = newNode.next;
      count += 1;
    }
    newNode.val = val;
    return true;
  }
  get(val) {
    if (val >= this.length) return undefined;
    var count = 0;
    var newNode = new Node();
    newNode = this.head;
    while (count != val) {
      newNode = newNode.next;
      count += 1;
    }
    return newNode.val;
  }
  insert(index, val) {
    if (index >= this.length) return null;
    if (index === 0) {
      this.unshift(val);
      return;
    }
    var count = 0;
    var curr = this.head;
    while (index - 1 != count) {
      curr = curr.next;
    }
    var newNode = new Node(val);
    newNode.next = curr.next;
    curr.next = newNode;
    this.length += 1;
  }
  remove(index) {
    if (index == this.length - 1) {
      this.pop();
      this.length -= 1;
      return;
    }
    if (index == 0) {
      this.shift();
      this.length -= 1;
      return;
    }
    var count = 0;
    var newNode = new Node();
    newNode = this.head;
    while (count != index - 1) {
      newNode = newNode.next;
      count += 1;
    }
    newNode.next = newNode.next.next;
    this.length -= 1;
    return;
  }
}
list = new SinglyLinkedList();
list.push(12);
list.push(13);
list.push(14);
list.push(10);
console.log(list);
list.pop();
console.log(list);
list.shift();
console.log(list);
list.unshift(11);
console.log(list);
console.log(list.get(2));
list.set(2, 990);
console.log(list);
list.insert(0, 12);
console.log(list);
list.remove(1);
console.log(list);
